File: src/PythonCode/attnscPV.py
'''
Sum check that simulates sending back and forth
'''
# %%
import numpy as np
import random
from math import log2
# %%
# ---------------------------
# 1) Configuration: Prime Field
# ---------------------------
P = 2**61 - 1
# %%
# ---------------------------
# 2) Hypercube Enumerator
# ---------------------------
def binary_vectors_lsb(b):
    """Yield each binary vector of length b, LSB-first."""
    for i in range(2**b):
        yield [(i >> k) & 1 for k in range(b)]
# %%
# ---------------------------
# 3) Multilinear Extension
# ---------------------------
def multilinear_extension_int(M_int):
    N, D = M_int.shape
    b_p = int(log2(N))
    b_q = int(log2(D))
    row_bits = list(binary_vectors_lsb(b_p))
    col_bits = list(binary_vectors_lsb(b_q))

    def mle(s, t):
        total = 0
        for i, ri in enumerate(row_bits):
            Li = 1
            for k in range(b_p):
                term = ((1 - s[k]) * (1 - ri[k]) + s[k] * ri[k]) % P
                Li = (Li * term) % P
            for j, cj in enumerate(col_bits):
                Lj = 1
                for m in range(b_q):
                    term = ((1 - t[m]) * (1 - cj[m]) + t[m] * cj[m]) % P
                    Lj = (Lj * term) % P
                total = (total + M_int[i, j] * Li * Lj) % P
        return total

    return mle
# %%
# ---------------------------
# 4) Build Attention Polynomial
# ---------------------------
def build_attention_F_int(X_int, Wq_int, Wk_int, i, j):
    _, e = X_int.shape
    _, d = Wq_int.shape
    b_p = int(log2(e))
    b_q = int(log2(d))

    mle_Xi = multilinear_extension_int(X_int[i:i+1, :].T)
    mle_Wq = multilinear_extension_int(Wq_int)
    mle_Xj = multilinear_extension_int(X_int[j:j+1, :d])
    mle_Wk = multilinear_extension_int(Wk_int[:d, :])

    num_bits = b_p + 2 * b_q

    def F(z_bits):
        s = z_bits[0:b_p]
        t = z_bits[b_p:b_p + b_q]
        u = z_bits[b_p + b_q:]
        return (mle_Xi(s, []) * mle_Wq(s, t) * mle_Xj([], t) * mle_Wk(t, u)) % P

    return num_bits, F
# %%
# ---------------------------
# 5) Prover Class
# ---------------------------# 
class Prover:
    def __init__(self, X_int, Wq_int, Wk_int, i, j):
        self.X = X_int
        self.Wq = Wq_int
        self.Wk = Wk_int
        self.i = i
        self.j = j
        
        # 3A) Compute direct raw attention score S_direct = Σ_{p,q,h} X_i,p Wq[p,q] X_j,q Wk[q,h] mod P
        self.S_direct = 0
        e = X_int.shape[1]
        d = Wq_int.shape[1]
        for p in range(e):
            for q in range(d):
                for h in range(d):
                    self.S_direct = (
                        self.S_direct
                        + (self.X[i,p] * self.Wq[p,q] % P)
                        * (self.X[j,q] * self.Wk[q,h] % P)
                    ) % P

        # 3B) Build the sum-check polynomial F(z) and record its dimension m
        self.num_bits, self.F = build_attention_F_int(X_int, Wq_int, Wk_int, i, j)

    def initial_claim(self):
        """Prover sends the raw attention score to the verifier."""
        return self.S_direct

    def send_partial_sums(self, challenges, r):
        """
        In round r, prover computes h_r(0) and h_r(1):
        h_r(u) = Σ_{tail ∈ {0,1}^{m-r-1}} F(challenges + [u] + tail)
        """
        def h(u):
            total = 0
            for tail in binary_vectors_lsb(self.num_bits - (r + 1)):
                z = challenges + [u] + tail
                total = (total + self.F(z)) % P
            return total

        return h(0), h(1)

    def compute_claim_at(self, challenges, r, u):
        """Helper to compute h_r(u) for new challenge u."""
        total = 0
        for tail in binary_vectors_lsb(self.num_bits - (r + 1)):
            z = challenges + [u] + tail
            total = (total + self.F(z)) % P
        return total

    def final_evaluation(self, challenges):
        """Prover sends F(challenges) at the very end."""
        return self.F(challenges) % P
# %%
# ---------------------------
# 4) Simulate Interaction
# ---------------------------
def simulate_prover_verifier(X_int, Wq_int, Wk_int, i, j):
    random.seed(0)
    prover = Prover(X_int, Wq_int, Wk_int, i, j)

    # Verifier receives initial claim
    claim = prover.initial_claim()
    print("Verifier receives initial claim S_direct =", claim)

    challenges = []
    # m rounds of sum-check
    for r in range(prover.num_bits):
        # Prover sends h(0), h(1)
        h0, h1 = prover.send_partial_sums(challenges, r)
        print(f"Round {r} • Prover → Verifier: h(0)={h0}, h(1)={h1}")

        # Verifier checks h0+h1 == claim
        assert (h0 + h1) % P == claim, f"Round {r} mismatch"

        # Verifier chooses challenge and sends to prover
        r_r = random.randrange(P)
        print(f"Round {r} • Verifier → Prover: challenge r_r={r_r}")

        # Prover computes new claim = h(r_r)
        new_claim = prover.compute_claim_at(challenges, r, r_r)
        challenges.append(r_r)
        claim = new_claim
        print(f"Round {r} • Updated claim = {claim}")

    # Final step: Verifier asks for F(challenges)
    final_val = prover.final_evaluation(challenges)
    print("Verifier receives final F(challenges) =", final_val)

    # Final consistency check
    assert final_val == claim, "Final check mismatch"
    print("Verifier accepts: attention score verified")
# %%
# ---------------------------
# 5) Demo
# ---------------------------
if __name__ == "__main__":
    # Toy data as before
    X = np.array([[1,2,3,4],[5,6,7,8],[9,10,11,12]],dtype=np.int64)
    Wq = np.array([[1,2],[3,4],[5,6],[7,8]],dtype=np.int64)
    Wk = np.array([[2,1],[4,3],[6,5],[8,7]],dtype=np.int64)

    X_int, Wq_int, Wk_int = X % P, Wq % P, Wk % P
    i, j = 1, 2

    print("▶ Simulating prover-verifier sum-check:")
    simulate_prover_verifier(X_int, Wq_int, Wk_int, i, j)

File: Outdated/CppCudaCode_outdated/attnscPV.cpp
#include <cstdint>
#include <vector>
#include <random>
#include <iostream>
#include <cassert>
#include <stdlib.h>
#include <algorithm>
#include <limits>
using namespace std;
using u64 = uint64_t;
static constexpr u64 P = (1ULL<<61) - 1;

// Compute a^e mod P in O(log e) time.
u64 modexp(u64 a, u64 e) {
    u64 res = 1;
    while(e) {
        if(e & 1) res = (u64)((__uint128_t)res * a % P);
        a = (u64)((__uint128_t)a * a % P);
        e >>= 1;
    }
    return res;
}

// We'll fill this at runtime once we know d:
u64 inv_sqrt_dk;
// fast field ops
inline u64 fadd(u64 a, u64 b) { u64 c = a+b; return c>=P?c-P:c; }
inline u64 fmul(u64 a, u64 b){ __uint128_t t=(__uint128_t)a*b; return (u64)(t%P); }

// hypercube enumerator
vector<vector<uint8_t>> binary_vectors_lsb(int b) {
    int N = 1<<b;
    vector<vector<uint8_t>> out(N, vector<uint8_t>(b));
    for(int i=0;i<N;i++)
      for(int k=0;k<b;k++)
        out[i][k] = (i>>k)&1;
    return out;
}

struct MLE {
    int N, D, b_p, b_q;
    const u64* M;                           // pointer to N×D data
    vector<vector<uint8_t>> row_bits;       
    vector<vector<uint8_t>> col_bits;

    // constructor: note semicolon after struct vars above
    MLE(const u64* M_data, int N_, int D_)
      : N(N_), D(D_), 
        b_p((int)log2(N_)), 
        b_q((int)log2(D_)),
        M(M_data)
    {
        row_bits = binary_vectors_lsb(b_p);
        col_bits = binary_vectors_lsb(b_q);
    }

    // evaluate multilinear extension at (s,t)
    u64 eval(const u64* s, const u64* t) const {
        u64 total = 0;
        for(int i=0;i<N;i++){
            // compute L_i(s)
            u64 Li = 1;
            for(int k=0;k<b_p;k++){
                u64 bit = row_bits[i][k];
                u64 term = fadd(
                  fmul(fadd(1, P - s[k]), (u64)(1 - bit)),
                  fmul(s[k], bit)
                );
                Li = fmul(Li, term);
            }
            // loop over columns
            for(int j=0;j<D;j++){
                u64 Lj = 1;
                for(int m=0;m<b_q;m++){
                    u64 bit = col_bits[j][m];
                    u64 term = fadd(
                      fmul(fadd(1, P - t[m]), (u64)(1 - bit)),
                      fmul(t[m], bit)
                    );
                    Lj = fmul(Lj, term);
                }
                // accumulate
                total = fadd(total,
                             fmul(M[i*D + j], fmul(Li, Lj)));
            }
        }
        return total;
    }
};

struct AttentionPoly {
    int b_p, b_q, m;
    MLE mle_Xi, mle_Wq, mle_Xj, mle_Wk;

    AttentionPoly(const u64 *X_int, const u64 *Wq_int, const u64 *Wk_int,
                  int n, int e, int d, int i, int j)
      : b_p(log2(e)), b_q(log2(d)), m(b_p + 2*b_q),
        mle_Xi(&X_int[i*e], e, 1),      // X_i row: e×1
        mle_Wq(Wq_int, e, d),           // Wq: e×d
        mle_Xj(&X_int[j*e], 1, d),      // X_j row: 1×d
        mle_Wk(Wk_int, d, d)            // Wk: d×d
    {}

    // Evaluate F(z_bits) mod P
    u64 eval(const vector<u64> &z) const {
        const u64 *s = z.data();
        const u64 *t = z.data() + b_p;
        const u64 *u = z.data() + b_p + b_q;
        u64 v1 = mle_Xi.eval(s, nullptr);
        u64 v2 = mle_Wq.eval(s, t);
        u64 v3 = mle_Xj.eval(nullptr, t);
        u64 v4 = mle_Wk.eval(t, u);
        return fmul(fmul(v1, v2), fmul(v3, v4));
    }
};

class Prover {
  public:
    u64 S_direct;
    AttentionPoly poly;
    int m;

    Prover(const u64* X, const u64* Wq, const u64* Wk,
           int n, int e, int d, int i, int j)
      : poly(X,Wq,Wk,n,e,d,i,j),
        m(poly.m)
    {
        // compute S_direct by brute‐force triple loop
        S_direct = 0;
        for(int p=0; p<e; p++)
          for(int q=0; q<d; q++)
            for(int h=0; h<d; h++){
              u64 term = fmul(X[i*e + p], Wq[p*d + q]);
              term = fmul(term, fmul(X[j*e + q], Wk[q*d + h]));
              S_direct = fadd(S_direct, term);
            }
    }

    // round‐0: send initial claim
    u64 initial_claim() const {
        return S_direct;
    }

    // rounds 1..m: send h_r(0), h_r(1)
    pair<u64,u64> send_partial_sums(const vector<u64>& challenges, int r) const {
        auto h = [&](u64 u)->u64{
            u64 tot = 0;
            int rem = m - (r+1);
            vector<u64> tail(rem), z2;
            for(int tix=0; tix<(1<<rem); tix++){
                z2 = challenges;
                z2.push_back(u);
                for(int k=0;k<rem;k++){
                    tail[k] = (tix>>k)&1;
                    z2.push_back(tail[k]);
                }
                tot = fadd(tot, poly.eval(z2));
            }
            return tot;
        };
        return { h(0), h(1) };
    }

    // after receiving challenge u, compute new claim = h_r(u)
    u64 compute_claim_at(const vector<u64>& challenges, int r, u64 u) const {
        auto sums = send_partial_sums(challenges, r);
        if(u==0) return sums.first;
        if(u==1) return sums.second;
        // for general u:
        u64 tot = 0;
        int rem = m - (r+1);
        vector<u64> tail(rem), z2;
        for(int tix=0; tix<(1<<rem); tix++){
            z2 = challenges;
            z2.push_back(u);
            for(int k=0;k<rem;k++){
                tail[k] = (tix>>k)&1;
                z2.push_back(tail[k]);
            }
            tot = fadd(tot, poly.eval(z2));
        }
        return tot;
    }

    // final: send F(challenges)
    u64 final_evaluation(const vector<u64>& challenges) const {
        return poly.eval(challenges);
    }
};

class Verifier {
  public:
    mt19937_64 rng;
    int m;

    Verifier(int m_) : rng(123), m(m_) {}

    bool run(Prover& prover) {
        // Round 0
        u64 claim = prover.initial_claim();
        cout<<"[Verifier] initial claim = "<<claim<<"\n";

        vector<u64> challenges;
        // Rounds 1..m
        for(int r=0; r<m; r++){
            auto [h0,h1] = prover.send_partial_sums(challenges, r); 
            cout<<"[Prover→Verifier] round "<<r<<": h(0)="<<h0<<", h(1)="<<h1<<"\n";
            if(fadd(h0,h1) != claim){
                cout<<"[Verifier] REJECT at round "<<r<<"\n";
                return false;
            }
          
            u64 rr = rng() % P;
            cout<<"[Verifier→Prover] challenge r_"<<r<<" = "<<rr<<"\n";
            
            u64 new_claim = prover.compute_claim_at(challenges, r, rr);
            challenges.push_back(rr);
            claim = new_claim;
            cout<<"[Prover→Verifier] new claim = "<<claim<<"\n";
        }

        // Final check
        u64 final_val = prover.final_evaluation(challenges);
        cout<<"[Verifier] final eval = "<<final_val<<", expected = "<<claim<<"\n";
        if(final_val != claim){
            cout<<"[Verifier] FINAL REJECT\n";
            return false;
        }
        cout<<"[Verifier] ACCEPT\n";
        return true;
    }
};

// ─── Demo in main() ─────────────────────────────────────────
int main(){
    // 1) Specify dimensions at runtime
    int n, e, d;
    cout << "Enter n (# tokens), e (embed dim), d (head dim): ";
    cin >> n >> e >> d;

    // 2) Allocate flat arrays as vectors
    vector<u64> X(n * e);
    vector<u64> Wq(e * d);
    vector<u64> Wk(e * d);  // same shape as Wq

    // 3) Fill them however you like—here, random small ints
    mt19937_64 rng(42);
    for(auto &x : X)  x = rng() % 10;  // values 0..9
    for(auto &w : Wq) w = rng() % 10;
    for(auto &w : Wk) w = rng() % 10;

    // 4) Pick an (i, j) pair to verify
    int i = rng() % n;
    int j = rng() % n;
    cout << "Verifying entry (i="<<i<<", j="<<j<<")\n";

    // 5) Construct prover and verifier
    Prover prover(X.data(), Wq.data(), Wk.data(), n, e, d, i, j);
    Verifier verifier(prover.m);

    // 6) Run the protocol
    cout << "Running sum-check …\n";
    bool ok = verifier.run(prover);
    cout << (ok ? "SUCCESS" : "FAIL") << "\n";

    return ok ? 0 : 1;
}
